### PR TITLE
Bug 1803589 - Implement submit site support requests for cookie banners in Focus

### PR DESCRIPTION
**Pull request by iorgamgabriel**
### Pull Request checklist

- [ ] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/android-components/blob/main/docs/changelog.md) or does not need one
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/main/android/accessibility_guide.md) or does not include any user facing features

### After merge
- **Milestone**: Make sure issues closed by this pull request are added to the [milestone](https://github.com/mozilla-mobile/android-components/milestones) of the version currently in development.
- **Breaking Changes**: If this is a breaking change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### GitHub Automation
https://bugzilla.mozilla.org/show_bug.cgi?id=1803589

### CODE REVIEW

File: focus-android/app/src/main/java/org/mozilla/focus/cookiebannerexception/CookieBannerExceptionStatus.kt [hunk 1]
```diff
@@ -0,0 +1,11 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package org.mozilla.focus.cookiebannerexception
+
+sealed class CookieBannerExceptionStatus {
```
Comment (Amejia481) [inline focus-android/app/src/main/java/org/mozilla/focus/cookiebannerexception/CookieBannerExceptionStatus.kt:7]: nit: docs :)

File: focus-android/app/src/main/java/org/mozilla/focus/settings/privacy/TrackingProtectionPanel.kt [hunk 2]
```diff
@@ -183,11 +194,17 @@ class TrackingProtectionPanel(
...
+                )
             }
 
             analytics.onClickListener {
-                updateTrackingProtectionPolicy(TrackerChanged.ANALYTICS.tracker, analytics.isChecked)
+                updateTrackingProtectionPolicy(
+                    TrackerChanged.ANALYTICS.tracker,
+                    analytics.isChecked,
+                )
             }
```
Comment (Amejia481) [inline focus-android/app/src/main/java/org/mozilla/focus/settings/privacy/TrackingProtectionPanel.kt:208]: nit: Let's try to only change selections related to the patch, it makes a it hard to see which changes are link to the ticket or just unrelated formatting changes  😄

**Comment by Amejia481** [comment]
This looks good in general 👍🏽 !

File: focus-android/app/src/main/java/org/mozilla/focus/settings/privacy/TrackingProtectionPanel.kt (hunk 2 above)
Comment (iorgamgabriel) [reply focus-android/app/src/main/java/org/mozilla/focus/settings/privacy/TrackingProtectionPanel.kt]: Done

File: focus-android/app/src/main/java/org/mozilla/focus/cookiebannerexception/CookieBannerExceptionStatus.kt (hunk 1 above)
Comment (iorgamgabriel) [reply focus-android/app/src/main/java/org/mozilla/focus/cookiebannerexception/CookieBannerExceptionStatus.kt:11]: Done

**Review by Amejia481 — APPROVED** [review]

File: focus-android/app/src/main/java/org/mozilla/focus/cookiebannerreducer/CookieBannerReducerMiddleware.kt [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+            context.store.dispatch(
+                CookieBannerReducerAction.ShowSnackBarForSiteToReport(
+                    true,
+                ),
+            )
+            context.store.dispatch(
+                CookieBannerReducerAction.UpdateCookieBannerReducerStatus(
+                    CookieBannerReducerStatus.CookieBannerUnsupportedSiteRequestWasSubmitted,
+                ),
+            )
```
Comment (Amejia481) [inline focus-android/app/src/main/java/org/mozilla/focus/cookiebannerreducer/CookieBannerReducerMiddleware.kt:106]: I'm not familiar with pings, but I guess we would like to keep track of the reported sites, should we include the site as part of the ping?

Comment (iorgamgabriel) [reply focus-android/app/src/main/java/org/mozilla/focus/cookiebannerreducer/CookieBannerReducerMiddleware.kt:106]: This is an answer from @travis79 . He helped me with this issue. 
Yes, Glean can record all of those things, in fact, every Glean ["ping"](https://mozilla.github.io/glean/book/user/pings/index.html) already contains much of this information.
Timestamp:  Glean has a few different types of these, including timestamps built into every ping to indicate the beginning and end of the window of time from which the data in the ping was collected. If you need a specific timestamp for something, Based on the doc you linked, I think that the built-in Glean timestamps combined with the use of a "custom" ping should probably get you what you want.
Domain: This is definitely Category 3/4 (highly-sensitive) data, and Glean provides a [URL](https://mozilla.github.io/glean/book/reference/metrics/url.html) data type for this sort of data. On top of that, sending this in a custom ping will help to isolate it from other telemetry and allow for a custom data-retention policy for it. Finally, using a custom ping will allow you to not include the telemetry client_id, which can further help to protect the user's privacy by making it harder to link to other data we may collect.
Country: We collect this information when the ping is sent, automatically from a geo-lookup using the IP address before we discard the IP address to protect privacy. Unless you are prompting the user for this information and cannot rely on the automated geo information. In that case, using a String data type to collect this if it is user input would be advisable.
Language/Locale: Just like country, Glean collects this information from the device automatically based on the user's settings. If you need to record something other than the automatic locale that Glean collects, a String data type would also be advisable to use for this.
Here is an example of a custom ping defined in Fenix that does not send the telemetry client-id with the custom ping data: https://github.com/mozilla-mobile/firefox-android/blob/8cf9723866b8343423cc5ac6112319dab7cc6be2/fenix/app/pings.yaml#L34
And here is an example from Fenix's metrics.yaml file of a metric sent in that "topsites-impression" custom ping. It's a URL that would be similar to the domain you wish to capture: https://github.com/mozilla-mobile/firefox-android/blob/main/fenix/app/metrics.yaml#L5942

File: focus-android/app/src/main/java/org/mozilla/focus/cookiebannerreducer/CookieBannerReducerMiddleware.kt [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+                if (!hasException) {
+                    callHasCookieBannerRuleForSession(context)
+                } else {
+                    showExceptionStatus(context, true)
+                }
+            }
+        }
+    }
+
+    private fun callHasCookieBannerRuleForSession(
```
Comment (Amejia481) [inline focus-android/app/src/main/java/org/mozilla/focus/cookiebannerreducer/CookieBannerReducerMiddleware.kt:136]: nit: I think we need to improve the function to be more descriptive, to indicate what is doing without telling the name of the function that it's calling internally. Maybe `showUnsupportedSiteIfNeeded`.

File: focus-android/app/src/main/java/org/mozilla/focus/cookiebannerreducer/CookieBannerReducerMiddleware.kt (hunk 3 above)
Comment (Amejia481) [reply focus-android/app/src/main/java/org/mozilla/focus/cookiebannerreducer/CookieBannerReducerMiddleware.kt:106]: Thanks for the context, I think we could add @travis79 as reviewer to verify this implementation and for data review as this is Category 3/4 (highly-sensitive) data.

File: focus-android/app/metrics.yaml [hunk 5]
```diff
@@ -2356,7 +2356,44 @@ cookie_banner:
     metadata:
       tags:
         - Privacy&Security
-
+  report_site_domain:
+    type: url
+    description: |
+      A user can report a site when the cookie banner
+      reducer is not working from the cookie banner details panel.
```
Comment (chutten) [inline focus-android/app/metrics.yaml:2363]: A good metric description is specific about what values the metric can and cannot contain, in addition to what the data might mean. You've done a good job on the second half, but from the name it sounds as though this isn't a "site when the cookie banner reducer is not working", but is just the domain? Is that so?

**Review by chutten — CHANGES_REQUESTED** [review]
Needs data review, tests.

File: focus-android/app/metrics.yaml [hunk 6]
```diff
@@ -2356,7 +2356,44 @@ cookie_banner:
...
+    lifetime: ping
+    send_in_pings:
+      - cookie-banner-report-site
+    bugs:
+      - https://github.com/mozilla-mobile/fenix/issues/
```
[the remark this answers is not in the document]
Comment (iorgamgabriel) [reply focus-android/app/metrics.yaml:2368]: I will update the links after a colleague has done a data review . This is an initial commit to see if the implementation is what is needed.

File: focus-android/app/pings.yaml [hunk 7]
```diff
@@ -16,3 +16,15 @@ activation:
...
+  description: |
+    This ping is needed when the cookie banner reducer doesn't work on
+    a website, and the user wants to report the site.
+  include_client_id: false
+  bugs:
+    - https://github.com/mozilla-mobile/focus-android/issues/
+  data_reviews:
+    - https://github.com/mozilla-mobile/focus-android/issues/
+  notification_emails:
+    - android-probes@mozilla.com
```
[the remark this answers is not in the document]
Comment (travis79) [reply focus-android/app/pings.yaml:30]: If I understand correctly, you are submitting a URL that is reported by the user, but wish to keep that URL information separated from the client-id (that's good). 

If that is the case, then everything I see here is on the correct path for doing that: the URL is indeed sent in the new custom ping you defined and that custom ping is correctly configured to not include the client-id. 

I second @chutten's suggestions, but I think that this does everything I would recommend in this situation.

File: focus-android/app/src/main/java/org/mozilla/focus/cookiebannerreducer/CookieBannerReducerMiddleware.kt [hunk 8]
```diff
@@ -0,0 +1,203 @@
...
+        }
+    }
+
+    private fun reportSite(
+        action: CookieBannerReducerAction.RequestReportSite,
+        context: MiddlewareContext<CookieBannerReducerState, CookieBannerReducerAction>,
+    ) {
+        CookieBanner.reportSiteDomain.set(action.siteToReport)
+        CoroutineScope(Dispatchers.IO).launch {
+            Pings.cookieBannerReportSite.submit()
```
[the remark this answers is not in the document]
Comment (travis79) [reply focus-android/app/src/main/java/org/mozilla/focus/cookiebannerreducer/CookieBannerReducerMiddleware.kt:96]: I second this. Glean takes great care to perform all operations that might perform I/O either to the disk or to the network off of the main thread and it should generally be unnecessary to call any Glean API from a coroutine like this. I don't think that this hurts anything but it _could_ cause some weird out of ordering issues. Glean manages a serial execution dispatcher to ensure the ordering of things so this additional parallel scope on top of that could affect that determined ordering if there were multiple calls to `reportSite` happening concurrently.

File: focus-android/app/src/main/java/org/mozilla/focus/cookiebannerreducer/CookieBannerReducerMiddleware.kt (hunk 3 above)
Comment (travis79) [reply focus-android/app/src/main/java/org/mozilla/focus/cookiebannerreducer/CookieBannerReducerMiddleware.kt:106]: The implementation looks good, separating the url data from any identifiers by use of a custom ping is preferred when sending something that could be sensitive like a URL, even if that url was submitted explicitly by the user.

**Review by travis79 — APPROVED** [review]

File: focus-android/app/src/main/java/org/mozilla/focus/cookiebannerreducer/CookieBannerReducerMiddleware.kt (hunk 4 above)
Comment (iorgamgabriel) [reply focus-android/app/src/main/java/org/mozilla/focus/cookiebannerreducer/CookieBannerReducerMiddleware.kt]: Done.

File: focus-android/app/src/main/java/org/mozilla/focus/cookiebannerreducer/CookieBannerReducerMiddleware.kt (hunk 8 above)
[the remark this answers is not in the document]
Comment (iorgamgabriel) [reply focus-android/app/src/main/java/org/mozilla/focus/cookiebannerreducer/CookieBannerReducerMiddleware.kt]: Done , I removed that Coroutine .

**Comment by iorgamgabriel** [comment]
@Delphine Could you please do a review of the strings file before I merge this pull request? Thank you very much !

File: focus-android/app/metrics.yaml (hunk 5 above)
Comment (iorgamgabriel) [reply focus-android/app/metrics.yaml]: Done , added site domain to the description.

File: focus-android/app/pings.yaml [hunk 9]
```diff
@@ -16,3 +16,15 @@ activation:
     - https://github.com/mozilla-mobile/focus-android/issues/4901
   notification_emails:
     - jalmeida@mozilla.com
+
+cookie-banner-report-site:
+  description: |
+    This ping is needed when the cookie banner reducer doesn't work on
+    a website, and the user wants to report the site.
```
[the remark this answers is not in the document]
Comment (iorgamgabriel) [reply focus-android/app/pings.yaml:23]: Done

**Comment by iorgamgabriel** [comment]
# Request for data collection review form

All questions are mandatory. You must receive review from a data steward peer on your responses to these questions before shipping new data collection.

1. What questions will you answer with this data?
  * How users interact with reporting a site domain when the cookie banner reducer is not working.
2. Why does Mozilla need to answer these questions? Are there benefits for users? Do we need this information to address product or business requirements? 
  * This data will help us understand how the reporting a site domain feature is used.
3. What alternative methods did you consider to answer these questions? Why were they not sufficient?
  * There are no other alternatives. 
4. Can current instrumentation answer these questions?
  * No. 
5. List all proposed measurements and indicate the category of data collection for each measurement, using the Firefox data collection categories found on the Mozilla wiki. 
Note that the data steward reviewing your request will characterize your data collection based on the highest (and most sensitive) category.

Measurement Description | Data Collection Category | Tracking Bug #
-- | -- | --
The user has reported a site domain  | Category 2 - interaction data |https://bugzilla.mozilla.org/show_bug.cgi?id=1803589
The user has pressed the report site domain button  | Category 2 - interaction data |https://bugzilla.mozilla.org/show_bug.cgi?id=1803589

6. Please provide a link to the documentation for this data collection which describes the ultimate data set in a public, complete, and accurate way.
  * https://dictionary.telemetry.mozilla.org/apps/focus_android
7. How long will this data be collected? Choose one of the following:
  * Until version 119. Renewal then may be decided.
8. What populations will you measure?
  * All channels, all locales, all countries  that the feature is active for.
9. If this data collection is default on, what is the opt-out mechanism for users?
  * Default Glean SDK opt-out mechanism. 
10. Please provide a general description of how you will analyze this data.
  * Glean and Amplitude. 
11. Where do you intend to share the results of your analysis?
  *  Only on Glean, Amplitude, and with mobile teams. 
12. Is there a third-party tool (i.e. not Telemetry) that you are proposing to use for this data collection?
   * No third-party tools.

**Comment by iorgamgabriel** [comment]
@travis79 Can you help me please with a data review ? Thank you very much .

**Review by travis79 — APPROVED** [review]
> # Request for data collection review form
> 
> All questions are mandatory. You must receive review from a data steward peer on your responses to these questions before shipping new data collection.
> 
>     1. What questions will you answer with this data?
> 
> 
>     * How users interact with reporting a site domain when the cookie banner reducer is not working.
> 
> 
>     2. Why does Mozilla need to answer these questions? Are there benefits for users? Do we need this information to address product or business requirements?
> 
> 
>     * This data will help us understand how the reporting a site domain feature is used.
> 
> 
>     3. What alternative methods did you consider to answer these questions? Why were they not sufficient?
> 
> 
>     * There are no other alternatives.
> 
> 
>     4. Can current instrumentation answer these questions?
> 
> 
>     * No.
> 
> 
>     5. List all proposed measurements and indicate the category of data collection for each measurement, using the Firefox data collection categories found on the Mozilla wiki.    
>        Note that the data steward reviewing your request will characterize your data collection based on the highest (and most sensitive) category.
> 
> 
> Measurement Description 	Data Collection Category 	Tracking Bug #
> The user has reported a site domain 	Category 2 - url 	[bugzilla.mozilla.org/show_bug.cgi?id=1803589](https://bugzilla.mozilla.org/show_bug.cgi?id=1803589)
> The user has pressed the report site domain cancel button from the cookie banner reducer details panel. 	Category 2 - interaction data 	[bugzilla.mozilla.org/show_bug.cgi?id=1803589](https://bugzilla.mozilla.org/show_bug.cgi?id=1803589)
> The user has pressed the report site domain button from the cookie banner reducer details panel. 	Category 2 - interaction data 	[bugzilla.mozilla.org/show_bug.cgi?id=1803589](https://bugzilla.mozilla.org/show_bug.cgi?id=1803589)
> 
>     6. Please provide a link to the documentation for this data collection which describes the ultimate data set in a public, complete, and accurate way.
> 
> 
>     * [dictionary.telemetry.mozilla.org/apps/focus_android](https://dictionary.telemetry.mozilla.org/apps/focus_android)
> 
> 
>     7. How long will this data be collected? Choose one of the following:
> 
> 
>     * Until version 119. Renewal then may be decided.
> 
> 
>     8. What populations will you measure?
> 
> 
>     * All channels, all locales, all countries    that the feature is active for.
> 
> 
>     9. If this data collection is default on, what is the opt-out mechanism for users?
> 
> 
>     * Default Glean SDK opt-out mechanism.
> 
> 
>     10. Please provide a general description of how you will analyze this data.
> 
> 
>     * Glean and Amplitude.
> 
> 
>     11. Where do you intend to share the results of your analysis?
> 
> 
>     * Only on Glean, Amplitude, and with mobile teams.
> 
> 
>     12. Is there a third-party tool (i.e. not Telemetry) that you are proposing to use for this data collection?
> 
> 
>     * No third-party tools.

# Data Review

1) Is there or will there be **documentation** that describes the schema for the ultimate data set in a public, complete, and accurate way?

Yes, through the metrics.yaml file and the [Glean Dictionary](https://dictionary.telemetry.mozilla.org/).

2) Is there a control mechanism that allows the user to turn the data collection on and off?

Yes, through the "Send Usage Data" preference in the application settings.

3) If the request is for permanent data collection, is there someone who will monitor the data over time?

N/A, collection to end or be extended by version 119

4) Using the **[category system of data types](https://wiki.mozilla.org/Data_Collection)** on the Mozilla wiki, what collection type of data do the requested measurements fall under?

Category 2. (The "reported_site_domain" may cross into Category 3, but it is _only_ sent if the user explicitly opts to send it to us.)

5) Is the data collection request for default-on or default-off?

Default-on, but I would like to add an additional note that the reported site domain is only sent if the user opts to send it to us, similar to a crash report.

6) Does the instrumentation include the addition of **any *new* identifiers** (whether anonymous or otherwise; e.g., username, random IDs, etc.  See the appendix for more details)?

No

7) Is the data collection covered by the existing Firefox privacy notice?

Yes

8) Does the data collection use a third-party collection tool?

No

## Result

data-review+

File: focus-android/app/pings.yaml [hunk 10]
```diff
@@ -16,3 +16,15 @@ activation:
...
+  include_client_id: false
+  bugs:
+    - https://github.com/mozilla-mobile/focus-android/issues/
+  data_reviews:
+    - https://github.com/mozilla-mobile/focus-android/issues/
```
[the remark this answers is not in the document]
Comment (iorgamgabriel) [reply focus-android/app/pings.yaml]: Done

File: focus-android/app/metrics.yaml (hunk 6 above)
[the remark this answers is not in the document]
Comment (iorgamgabriel) [reply focus-android/app/metrics.yaml]: Done

**Review by Delphine — APPROVED** [review]
LGTM!

File: focus-android/app/metrics.yaml [hunk 11]
```diff
@@ -2357,6 +2360,64 @@ cookie_banner:
       tags:
         - Privacy&Security
```
Comment (chutten) [inline focus-android/app/metrics.yaml:2362]: I suppose now's not the best time to handle this... but you may wish to consider filing a follow-up to split this file by component. May make it easier to maintain, and then you can use the file-level `$tags` instead of annotating each and every metric definition.

File: focus-android/app/metrics.yaml [hunk 12]
```diff
@@ -2357,6 +2360,64 @@ cookie_banner:
...
 
+  report_site_domain:
+    type: url
+    description: |
+      A user can report a site domain when the cookie banner
+      reducer is not working from the cookie banner details panel.
```
Comment (chutten) [inline focus-android/app/metrics.yaml:2367]: I recommend explaining what is meant by "domain" here. It appears to be eTLD + 1 (if I'm reading this Kotlin properly)

**Review by chutten — APPROVED** [review]
This looks good as-is, though I have a couple of notes.

Also: does the instrumentation have tests? Or is it expected to be included in some manual testing suites?

File: focus-android/app/metrics.yaml (hunk 12 above)
Comment (iorgamgabriel) [reply focus-android/app/metrics.yaml:2367]: Done

**Comment by iorgamgabriel** [comment]
@chutten I made a task for that https://bugzilla.mozilla.org/show_bug.cgi?id=1823611
